### PR TITLE
Introduce an `HistogramBase` interface in the GleanSDK

### PR DESCRIPTION
**Pull request by Dexterp37**
The context for this is bug 1566368: this is needed in order for the glean parsers to generate an API to route batched streaming telemetry calls to the Glean SDK. It additionally:

* Changes the error reporting mechanism to allow for reporting more than 1 error of the same type.
* Adds support for accumulating multiple samples in `TimingDistributionStorageEngine`

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.

### CODE REVIEW

File: components/service/glean/src/main/java/mozilla/components/service/glean/error/ErrorRecording.kt [hunk 1]
```diff
@@ -41,12 +41,15 @@ object ErrorRecording {
      *     ping.  It does not need to include the metric name, as that is automatically
      *     prepended to the message.
      * @param logger The [Logger] instance to display the warning.
+     * @param numErrors The optional number of errors to report for this [ErrorType].
      */
+    @Suppress("LongMethod")
```
Comment (travis79) [inline components/service/glean/src/main/java/mozilla/components/service/glean/error/ErrorRecording.kt]: Wow, adding that little bit pushed it over the `LongMethod` limit?

Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/error/ErrorRecording.kt]: Yes :| That was a bit surprising to me too, but... meh :|

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt [hunk 2]
```diff
@@ -97,6 +98,54 @@ internal open class TimingDistributionsStorageEngineImplementation(
...
+                currentValue?.let {
+                    it.accumulate(sample)
+                    it
+                } ?: let {
+                    val newTD = TimingDistributionData(category = metricData.category, name = metricData.name,
+                        timeUnit = timeUnit)
+                    newTD.accumulate(sample)
+                    return@let newTD
+                }
+            }
```
Comment (mdboom) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt]: Maybe reduce the code duplication between this and `accumulate` by just calling this from accumulate with an array of length 1?  I'm concerned about these two code paths becoming out-of-sync in the future.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt [hunk 3]
```diff
@@ -124,7 +173,8 @@ internal open class TimingDistributionsStorageEngineImplementation(
  * @param category of the metric
  * @param name of the metric
  * @param bucketCount total number of buckets
- * @param range an array always containing 2 elements: the minimum and maximum bucket values
+ * @param rangeMin the minimum value that can be represented
+ * @param rangeMin the maximum value that can be represented
```
Comment (mdboom) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt]:
```suggestion
 * @param rangeMax the maximum value that can be represented
```

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt (hunk 2 above)
Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt]: Good point. I changed this but I had to leave the negative value check in place. See the comment in the code.

**Pull request merged**